File: src/main/java/io/nightfis/createstellarexoduscore/filter/TargetFilter.java
package io.nightfis.createstellarexoduscore.filter;

import net.minecraft.network.chat.Component;
import net.minecraft.world.entity.LivingEntity;

import java.util.Collection;
import java.util.HashMap;

public abstract class TargetFilter {

    private static final TargetFilter EMPTY_FILTER = new TargetFilter() {
        @Override
        public String getId() {
            return "";
        }

        @Override
        public boolean match(LivingEntity entity, String arg) {
            return false;
        }

        @Override
        public Component title() {
            return Component.translatable("screen.createstellarexoduscore.auto_aim_turret.filter.unknown");
        }
    };

    private static final HashMap<String, TargetFilter> FILTER_MAP = new HashMap<>();

    public static void register(TargetFilter filter) {
        FILTER_MAP.put(filter.getId(), filter);
    }

    public static TargetFilter getFilter(String id) {
        return FILTER_MAP.getOrDefault(id, EMPTY_FILTER);
    }

    public static Collection<TargetFilter> all() {
        return FILTER_MAP.values();
    }

    public abstract String getId();

    public abstract boolean match(LivingEntity entity, String arg);

    public boolean needArg() {
        return false;
    }

    public abstract Component title();

    public String tooltip(String arg) {
        return "";
    }

    public static void init() {
        FILTER_MAP.clear();
        register(new IsPlayerFilter());
        register(new IsPlayerIdFilter());
        register(new IsPetFilter());
        register(new IsOnVehicleFilter());
        register(new IsFriendlyFilter());
        register(new IsEntityIdFilter());
        register(new IsBabyFriendlyFilter());
        register(new HasCustomNameFilter());
    }
}

File: src/main/kotlin/io/nightfis/createstellarexoduscore/filter/TargetFilter.kt
package io.nightfis.createstellarexoduscore.filter

import net.minecraft.network.chat.Component
import net.minecraft.world.entity.LivingEntity

abstract class TargetFilter {

    abstract fun getId(): String

    abstract fun match(entity: LivingEntity, arg: String): Boolean

    open fun needArg(): Boolean = false

    abstract fun title(): Component

    open fun tooltip(arg: String): String = ""

    companion object {
        private val EMPTY_FILTER: TargetFilter = object : TargetFilter() {
            override fun getId(): String = ""

            override fun match(entity: LivingEntity, arg: String): Boolean = false

            override fun title(): Component {
                return Component.translatable("screen.create_stellar_exodus_core.auto_aim_turret.filter.unknown")
            }
        }

        private val FILTER_MAP = HashMap<String, TargetFilter>()

        @JvmStatic
        fun register(filter: TargetFilter) {
            FILTER_MAP[filter.getId()] = filter
        }

        @JvmStatic
        fun getFilter(id: String): TargetFilter {
            return FILTER_MAP[id] ?: EMPTY_FILTER
        }

        @JvmStatic
        fun all(): Collection<TargetFilter> = FILTER_MAP.values

        @JvmStatic
        fun init() {
            FILTER_MAP.clear()
            register(IsPlayerFilter())
            register(IsPlayerIdFilter())
            register(IsPetFilter())
            register(IsOnVehicleFilter())
            register(IsFriendlyFilter())
            register(IsEntityIdFilter())
            register(IsBabyFriendlyFilter())
            register(HasCustomNameFilter())
        }
    }
}
